python/pyarrow/array.pxi

```diff
@@ -1210,20 +1218,22 @@ cdef class Array(_PandasConvertible):
         If a buffer is referenced multiple times then it will
         only be counted once.
         """
-        cdef:
-            int64_t total_buffer_size
-
+        self._assert_cpu()
+        cdef int64_t total_buffer_size
         total_buffer_size = TotalBufferSize(deref(self.ap))
```

Comment:
My first thought was that getting the buffer size should be possible without looking at the actual data. But so it seems that to avoid counting identical buffers twice (if they are reused in a single array, which is possible), it uses the buffer's address to distinguish buffers. Currently that uses `buffer->data()` in `DoTotalBufferSize`. `data()` will return null for non-cpu data, but since it doesn't actually use that address, I think this could also use `buffer->address()` which will return the address always even for non-cpu data.

(but that could be fine for a follow-up as well)

Reply:
That checks out! I tested it locally and witnessed a segfault, but hadn't dug into the reasoning.